src/Tools/DynamoInstallDetective/ProductLookUp.cs

```diff
@@ -9,6 +9,118 @@
+            {
+                DisableCache();
+            }
+        }
+
+        public static string REG_KEY64 = @"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\";
+        public static string REG_KEY32 = @"SOFTWARE\Wow6432Node\Microsoft\Windows\CurrentVersion\Uninstall\";
+
+        // Cached data from windows registry. Consists of a dictionary that maps The product name/code to Display Name (Tuple.item1) and Install Path(Tuple.item2)
+        // Dictionary<ProductCode, (DisplayName, InstallPath)>
+        private static Dictionary<string, Tuple<string, string>> cachedRecords;
+        private static bool cacheEnabled = false;
+
+        private static readonly object mutex = new object();
+
+        public static RegistryKey OpenKey(string key)
+        {
+            var regKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64);
+            return regKey.OpenSubKey(key);
+        }
+        public static string GetInstallLocation(RegistryKey key)
+        {
+            if (key != null)
+                return key.GetValue("InstallLocation") as string;
+
+            return string.Empty;
+        }
+
+        public static string GetDisplayName(RegistryKey key)
+        {
+            if (key != null)
+                return key.GetValue("DisplayName") as string;
+
+            return string.Empty;
+        }
+
+        // Returns all the products registered under "SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\" that have a valid DisplayName and an InstallLocation.
+        public static Dictionary<string, Tuple<string, string>> GetInstalledProducts()
+        {
+            lock (mutex)
+            {
+                if (cacheEnabled && cachedRecords != null)
+                {
+                    return cachedRecords;
+                }
+            }
+
+            var productInfo = new Dictionary<string, Tuple<string, string>>();
+
+            var key = OpenKey(REG_KEY64);
+            foreach(string s in key.GetSubKeyNames())
+            {
+                try
+                {
+                    var prodKey = key.OpenSubKey(s);
+                    var displayName = GetDisplayName(prodKey);
+                    var installLocation = GetInstallLocation(prodKey);
+                    if (displayName.Length > 0 && installLocation.Length > 0)
+                    {
+                        productInfo.Add(s, Tuple.Create(displayName, installLocation));
+                    }
+                }
+                catch (Exception)
+                { }
+            }
+
+            lock (mutex)
+            {
+                if (cacheEnabled)
+                {
+                    cachedRecords = new Dictionary<string, Tuple<string, string>>(productInfo);
+                }
+            }
+            return productInfo;
+        }
+
+        public static IDisposable StartCache()
+        {
+            return new RegistryCacher();
+        }
+
+        public static void EnableCache()
```

Comment:
Why not make `EnableCache`, `DisableCache`, and `cacheEnabled` private members of `RegistryCacher`? I think `RegistryCacher` should also have access to private members of `RegUtils`?

Reply:
Even those these are public - the class is internal, so these are really internal methods AFAIK.

Reply:
I guess that would work.
The only concern is that it would make the code harder to understand ...
The only purpose of the RegistryCacher class was to make it easier to enable/disable caching (by using a Disposable object)

Reply:
The fact that the `RegistryCacher` deals with caching, I think it makes sense to do anything related to caching part of the same class, meaning including `EnableCache` and `DisableCache`.

Reply:
In any case, I don't think it makes sense to have `Enable/DisableCache` public.

Reply:
actually that is a good point. we don't really need the EnableCache, DisableCache methods at the moment.
With that in mind it would make sense to move the inner logic to the RegistryCacher